### PR TITLE
[libc++] Correctly handle custom deleters in hardened unique_ptr<T[]>

### LINKED ISSUE #110683: [libc++] Hardened unique_ptr<T[]> incorrectly handles custom deleters
> This makes the attached program crash (…in most runs, not 100% of the time):
>
> [unique.cc.zip](https://github.com/user-attachments/files/17196807/unique.cc.zip)
>
> ```
>  % out/gn/bin/clang++ unique.cc -isysroot $(xcrun -show-sdk-path) -std=c++20 -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE     
> % ./a.out
> zsh: trace trap  ./a.out
> ```
> 
> lldb shows that this indeed this check:
> 
> ```
>     frame #0: 0x00000001000029e4 a.out`std::__1::unique_ptr<std::__1::vector<int, std::__1::allocator<int>> [], AlignedDeleter<std::__1::vector<int, std::__1::allocator<int>>>>::operator[][abi:se200000](unsigned long) const [inlined] __clang_trap_msg$libc++$/Users/thakis/src/llvm-project/out/gn/bin/../include/c++/v1/__memory/unique_ptr.h:566: assertion __checker_.__in_bounds(std::__to_address(__ptr_), __i) failed: unique_ptr<T[]>::operator[](index): index out of range
> ```
> 
> The program uses a unique_ptr with a custom deleter to refer to raw malloc-ed memory, "to allow having uninitialized entries inside it".
> 
> It has this typedef:
> 
> ```
> template <typename T>
> using AlignedUniquePtr =
>     std::unique_ptr<T, AlignedDeleter<typename std::remove_extent<T>::type>>;
> ```
> 
> It uses this typedef with an array type:
> 
> ```
>   // Underlying storage. It's raw malloc-ed rather than being a unique_ptr<T[]>
>   // to allow having uninitialized entries inside it.
>   AlignedUniquePtr<T[]> entries_;
> ```
> 
> It creates objects of this unique_ptr like so:
> 
> ```
> template <typename T>
> AlignedUniquePtr<T> AlignedAllocTyped(size_t n_membs) {
>   using TU = typename std::remove_extent<T>::type;
>   return AlignedUniquePtr<T>(
>       static_cast<TU*>(AlignedAlloc(alignof(TU), sizeof(TU) * n_membs)));
> }
> ```
> 
> `AlignedAlloc()` is basically malloc – in particular, nothing calls `new[]`, so it doesn't write an array cookie.
> 
> The custom deleter looks like so (`AlignedFree()` is just `free()`):
> 
> ```
> template <typename T>
> struct AlignedDeleter {
>   inline void operator()(T* ptr) const { AlignedFree(ptr); }
> };
> ```
> 
> The class containing `entries_` calls dtors as needed, and the deleter really only needs to free memory. `delete []` is never called.
> 
> Is this a valid use of unique_ptr<T[]>? If so, should this hardening only be used for unique_ptrs that don't have a custom deleter?

_Originally posted by @nico in https://github.com/llvm/llvm-project/issues/91798#issuecomment-2384260976_

### PR DESCRIPTION
**Pull request by ldionne**
It turns out that we can never do bounds-checking for unique_ptrs with custom deleters, except when converting from a unique_ptr with a default deleter to one with a custom deleter.

If we had an API like `std::make_unique` that allowed passing a custom deleter, we could at least get bounds checking when the unique_ptr is created through those APIs, but for now that is not possible.

Fixes #110683

### CODE REVIEW

File: libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/op_subscript.runtime.pass.cpp [hunk 1]
```diff
@@ -112,12 +117,33 @@ TEST_CONSTEXPR_CXX23 bool test() {
...
+        assert(p[1] == T());
+        assert(p[2] == T());
+      }
+
+      // Array not allocated with `new T[n]`, custom deleter
+      //
+      // This test aims to ensure that the implementation doesn't try to use an array cookie
+      // when there is none.
+      {
+        T array[50] = {};
```
Comment (ldionne) [inline libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/op_subscript.runtime.pass.cpp:141]: @nico I verified locally and this test breaks down without my patch. That's a smaller version of your reproducer.

**Review by nico — APPROVED** [review]
Thanks for the quick fix!

Might be nice to give custom deleters some way to opt in to this, but that can be for the future :)

**Comment by ldionne** [comment]
> Thanks for the quick fix!
> 
> Might be nice to give custom deleters some way to opt in to this, but that can be for the future :)

I am iffy about exposing a vendor extension for this. Instead I think it would make more sense to pursue standardization of `make_unique`-like APIs that support a custom deleter.

**Comment by nico** [comment]
> I am iffy about exposing a vendor extension for this. Instead I think it would make more sense to pursue standardization of `make_unique`-like APIs that support a custom deleter.

How does would that API help? Would it call `new []` internally for `T[]`s and require the custom deleter to eventually call `delete[]`?

(To be clear I agree about the iffy and don't want to argue for anything here, just asking for my understanding.)

**Comment by ldionne** [comment]
> 
> How does would that API help? Would it call `new []` internally for `T[]`s and require the custom deleter to eventually call `delete[]`?

`std::make_unique` currently looks like

```
auto ptr = std::make_unique<T[]>(10); // allocated using new T[10]
```

We know the size of the allocation because it gets passed to the function. The same would be true with a custom deleter, e.g.

```
auto ptr = std::make_unique<T[]>(10, mydeleter);
```

If that were the case, we could stash the size passed to `make_unique` inside the `unique_ptr` when we have the right ABI configuration. However, this doesn't help the case where we don't have the "fat" ABI configuration at all.

**Comment by nico** [comment]
Looks like the two failing bots are true failures:

[stage2 (generic-cxx11, clang-19, clang++-19)](https://github.com/llvm/llvm-project/actions/runs/11128390753/job/31006541749?pr=110685#logs):

```
# executed command: /usr/bin/clang++-19 /home/runner/_work/llvm-project/llvm-project/libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/op_subscript.runtime.pass.cpp -pthread --target=x86_64-unknown-linux-gnu -nostdinc++ -I /home/runner/_work/llvm-project/llvm-project/build/generic-cxx11/libcxx/test-suite-install/include/c++/v1 -I /home/runner/_work/llvm-project/llvm-project/build/generic-cxx11/libcxx/test-suite-install/include/c++/v1 -I /home/runner/_work/llvm-project/llvm-project/libcxx/test/support -std=c++11 -Werror -Wall -Wctad-maybe-unsupported -Wextra -Wshadow -Wundef -Wunused-template -Wno-unused-command-line-argument -Wno-attributes -Wno-pessimizing-move -Wno-noexcept-type -Wno-atomic-alignment -Wno-reserved-module-identifier -Wdeprecated-copy -Wdeprecated-copy-dtor -Wno-user-defined-literals -Wno-tautological-compare -Wsign-compare -Wunused-variable -Wunused-parameter -Wunreachable-code -Wno-unused-local-typedef -Wno-local-type-template-args -Wno-c++11-extensions -Wno-unknown-pragmas -Wno-pass-failed -Wno-mismatched-new-delete -Wno-redundant-move -Wno-self-move -D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER -D_LIBCPP_ENABLE_EXPERIMENTAL -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_NONE -Werror=thread-safety -Wuser-defined-warnings -lc++experimental -nostdlib++ -L /home/runner/_work/llvm-project/llvm-project/build/generic-cxx11/libcxx/test-suite-install/lib -Wl,-rpath,/home/runner/_work/llvm-project/llvm-project/build/generic-cxx11/libcxx/test-suite-install/lib -lc++ -latomic -o /home/runner/_work/llvm-project/llvm-project/build/generic-cxx11/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/Output/op_subscript.runtime.pass.cpp.dir/t.tmp.exe
# .---command stderr------------
# | /home/runner/_work/llvm-project/llvm-project/libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/op_subscript.runtime.pass.cpp:51:23: error: constexpr function's return type 'void' is not a literal type
# |    51 |   TEST_CONSTEXPR void operator()(T*) const {}
# |       |                       ^
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1
```

[macos (apple-system-hardened, macos-13)](https://github.com/llvm/llvm-project/actions/runs/11128390753/job/31006543256?pr=110685#logs):

```
# executed command: /Users/runner/work/llvm-project/llvm-project/.venv/bin/python3.12 /Users/runner/work/llvm-project/llvm-project/libcxx/utils/run.py --execdir /Users/runner/work/llvm-project/llvm-project/build/apple-system-hardened/cxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/Output/assert.subscript.pass.cpp.dir -- /Users/runner/work/llvm-project/llvm-project/build/apple-system-hardened/cxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/Output/assert.subscript.pass.cpp.dir/t.tmp.exe
# .---command stderr------------
# | Failure: EXPECT_DEATH( other[6] ) failed!
# | (reason: <invalid cause (child did not die)>)
# | 
# | child exit code: 5
# | ---------- standard err ----------
# | 
# | ----------------------------------
# | ---------- standard out ----------
# | 
# | ----------------------------------
# | Assertion failed: (( ExpectDeath(DeathCause::Trap, "other[6]", [&]() { (void)(other[6]); }) )), function operator(), file assert.subscript.pass.cpp, line 120.
# `-----------------------------
# error: command failed with exit status: 2
```

**Pull request merged**